psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1600 @@
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "psi4/libfock/jk.h"
+
+#include "psi4/libmints/siminteri.h"
+#include "psi4/lib3index/3index.h"
+#include "psi4/lib3index/dftensor.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/twobody.h"
+#include "psi4/libmints/matrix.h"
+#include "psi4/pragma.h"
+#include "psi4/libqt/qt.h"
+
+#include <cstdlib>
+#include <memory>
+#include <vector>
+#include <algorithm>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#include "psi4/libpsio/psio.hpp"
+#include "psi4/libpsio/psio.h"
+#include "psi4/libpsio/aiohandler.h"
+#include <sstream>
+#include "psi4/libpsi4util/PsiOutStream.h"
+
+
+
+namespace psi {
+
+DirectDFJK::DirectDFJK(std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> auxiliary )
+    : JK(primary), auxiliary_(auxiliary)  {
+// Not much happens in common init, just that the number of elements in the
+//  basis sets gets saved as private member data in the DirectDFJK object.
+//  However, the JK() constructor gets called as well. Not to fear!
+//  In our case, all that does is set a couple booleans. 
+    common_init();
+}
+
+DirectDFJK::~DirectDFJK() {}
+
+//It's hard to say what does and doesn't belong in this function. I am
+// open to suggestions as to what should change. JSOB
+void DirectDFJK::common_init(){
+    nbf_ = primary_->nbf();
+    naux_ = auxiliary_->nbf();
+
+	p_shells_ = primary_->nshell();
+	Q_shells_ = auxiliary_->nshell();
+}
+
+//Used as testing machinery. It's nice to have a public function
+//to be able to look at private member data from pyside.
+//I plan on nuking this as soon as I get this project done,
+//don't you worry. JSOB
+
+//syntax copied and adapted from DiskDFJK.
+//I decided to forgo copying from MemDFJK because
+//that code seems more specialized. 
+void DirectDFJK::print_header() const {
+	if (print_)	{
+		outfile->Printf("==> DirectDFJK: Density-Fitted J/K Matrices <==\n\n");
+
+        outfile->Printf("    J tasked:          %11s\n", (do_J_ ? "Yes" : "No"));
+        outfile->Printf("    K tasked:          %11s\n", (do_K_ ? "Yes" : "No"));
+		outfile->Printf("    wK tasked:         %11s\n", (do_wK_ ? "Yes" : "No"));
+		if (do_wK_) outfile->Printf("Don't interpret these results.\nWe don't have w integrals in DirectDFJK\n");
```

Comment:
Why not?

Reply:
The formulation we chose for the Fock matrix components relies on a symmetry that gets broken in the DF approximation to the omega integrals. Adding this capability would represent a significant change and would be another PR in its own right.

Reply:
Which symmetry..?

Reply:
B^Q_{i \mu } = B^Q_{i \mu } following Weigend's notation

Reply:
That's an identity..?

Reply:
In the contraction

K^{\mu \nu} = B^Q_{i \mu} B^Q_{i \nu}

We need that identity to hold. In the omega eri case, that symmetry does not hold, so a different strategy than the one taken in this PR would need to be used.

Reply:
I *think* what's going on here is the handling of the metric inverse term is different.  The B notation is the three index tensor contracted against the inverse square root of the Coulomb metric.  It looks like the erfc terms are not handled this way; one of the three index tensors is just the Coulomb term, while the other is the Coulomb term contracted against the inverse erfc metric.  Both metrics should be positive semidefinite (I think) so it's not immediately obvious to me why this asymmetry is introduced.

Reply:
You're right. It wouldn't introduce the problem that I assumed it would. In any case, it would be a lot of work to extend the code to the omega case, and I graduated last week. 

When this code was written over a year ago, the omega extension wasn't seen as a high priority, so it wasn't included. I am trying to finish this PR in the next couple weeks, so in any case, I won't be able to add this feature.

Reply:
It also isn't immediately clear to me how to generalize the particular algorithm in this code to the omega case. Worst case scenario, we can ask the user to use disk.

Reply:
Congrats @obrien951 !

Reply:
Generalizing to the omega case should be as simple as replacing all the integrals with their attenuated versions.

Of course, then J and K have be evaluated separately.

Reply:
... which may be what one wants anyhow, since for large systems forming K with DF may be slower than with exact integrals.

Reply:
@susilehtola Thank you for your well wishes

Because of the particulars of my blocking scheme, extending to the omega case actually will require a significant degree of work both in formulating the working equations and writing the code.  Someone would have to write out the equations to figure which block of integrals would be attenuated and which block would be contracted against the coulomb metric, assuming we're still using the working equation from DiskDFJK:

<sup>&omega;</sup> K<sub>&mu;&nu;</sub> = C<sub>&lambda;i</sub>(Q|&mu;&lambda;) (Q|&omega;|&nu;&sigma;) C<sub>&sigma;i</sub>

Someone would have to rewrite a loop over the blocking index to compute blocks of  C<sub>&lambda;i</sub>(Q|&mu;&lambda;) over some values of the outer blocking index and (Q|&omega;|&nu;&sigma;) C<sub>&sigma;i</sub> over others. I chose Q for both tensors here although only the non-omega tensor is contracted against the inverse coulomb metric. They would also have to see if this is better than dealing with having to have more blocks over the auxiliary blocking index as would be required by the larger number of tensors when dealing with attenuated integrals.

Luckily, in my experience with working with attenuated integrals, the extra time to calculate a tensor (A|&omega;|&nu;&sigma;)  of those integrals rather than a tensor (A|&mu;&lambda;) (index changed to A because this time it's uncontracted) is not the thing that's going to make the code takes longer.

Furthermore, while our testing shows that blocking over the primary basis set is faster for the non-attenuated case, it is not clear that this will hold for the attenuated case which will require more testing and which is a reason that I kept the code for blocking over the auxiliary basis set in this PR. This is all without mentioning coding up combined calculation of exchange terms.

Both formulating and testing that code would represent a significant endeavor that would require large additions to the code included in this PR.

Reply:
I think all integrals should be attenuated, so then it really is as simple as replacing the integral call from `1/r12` to the range separated kernel `[a*erf(w r12) + b*erfc(w r12)/r12]`

Reply:
If I'm not mistaken, you run into problems because then you're applying the kernel twice in the formulation. Rather than once in the conventional (that is the four-index eri) case. That said, if there's a formulation of the working equations that works better, I might be able to work into it.

When I looked through the DiskDFJK code and when I wrote the MemDFJK wK code which is currently working, the working equations involved one tensor of coulomb-operator 3-index integrals contracted against the inverse coulomb metric and another tensor of attenuated integrals as in

<sup>&omega;</sup> K<sub>&mu;&nu;</sub> = C<sub>&lambda;i</sub>(Q|&mu;&lambda;) (Q|&omega;|&nu;&sigma;) C<sub>&sigma;i</sub>

Reply:
i.e. <sup>&omega;</sup> K<sub>&mu;&nu;</sub> = C<sub>&lambda;i</sub>(A|&mu;&lambda;)(A|Q)<sup>-1</sup> (Q|&omega;|&nu;&sigma;) C<sub>&sigma;i</sub>

Reply:
Um, I don't think so; it's just that you define the fitting in terms of the fitted kernel. Like you say, the range-separated exchange is
 <sup>&omega;</sup> K<sub>&mu;&nu;</sub> = &sum;<sub>i occ</sub> C<sub>&lambda;i</sub> (&mu;&lambda;|&omega;|&nu;&sigma;) C<sub>&sigma;i</sub>

Now you just do RI employing the &omega; metric as
(&mu;&lambda;|&omega;|&nu;&sigma;) &approx; (&mu;&lambda;|&omega;|A) (A|&omega;|B)<sup>-1</sup> (B|&omega;|&nu;&sigma;)
where your fit functions are normalized (A|&omega;|A) = 1.

This is exactly the same as the usual global hybrid: 
(&mu;&lambda;|&nu;&sigma;) &approx; (&mu;&lambda;|A) (A|B)<sup>-1</sup> (B|&nu;&sigma;)  with (A|A) = 1 if you just redefine how the integrals are calculated.

Reply:
I'll see how that goes. It would be great to expand the capability of the code.

Reply:
Yeah. I'm not sure where Psi4 normalizes the auxiliary functions. I think the code might work even if Coulomb normalization is used; the main thing is just that one does not want the norms to vary by orders of magnitude since then the eigensolver gets into problems. Also the eigenvalues will be affected; the cutoffs are always defined in terms of normalized matrices so that's another hiccup. (The code in `orthog.cc` handles this properly by normalizing the overlap matrix before getting its eigenvalues.)

Reply:
What do you mean when you say "normalize the auxiliary functions?" I haven't see that language to describe any of these processes before.

Reply:
But I can tell you for a fact that the working equations I wrote above are used in Psi because I'm the one who wrote that code.  I will, however, make myself look foolish if I go around telling people that that's the only way to do it. I haven't followed up on your working equation because I'm still working on cleaning up the rest of the code.

Reply:
If you look up the literature on RI, they don't use the overlap normalization `<A|A>=1` but rather the Coulomb normalization `(A|A)=1`.

Reply:
Sorry for taking so long. I had to break for the holiday.

I tried out your attenuated coulomb metric, but I had some serious issues with conditioning. Psi4 actually has software written to help deal with these issues, and I used them in my trial. My code has the steps in the proper order, but when I try to take the half-power coulomb metric, either DSYEV fails or the energies are completely wrong. I read the Ahlrichs paper. He gives some strategies for dealing with some of the conditioning and all those issues. Implementing them might best be done in the lib3index/fittingmetric.cc, and would likely require a separate PR.

Furthermore, Head-Gordon has shown that the coulomb metric is able to meet many of the needs sought by attenuated metrics 

https://doi.org/10.1073/pnas.0408475102

and I am told that the Coulomb metric is seen as a more conventional way of performing density fitted attenuated calculations largely for this reason.

I think what I'm failing to communicate here is that if the two tensors between the coefficient matrices are not the same then the code must be rewritten starting with a new algorithm. I think that the amount of work that will go into extending direct-df calculations to attenuated integrals in a way that will not be restricted by system size is enough for a separate PR, and it would not have been possible for me to do in the time frame that came about from when we reran the trials of this code at the beginning of this fall semester.

I really hope I'm not coming off as recalcitrant or lazy. I tried the attenuated metric, and it didn't work within the parameters I needed it to.

Reply:
OK. Kállay also suggests using the Coulomb metric in https://pubs.acs.org/doi/10.1021/acs.jctc.9b00891, as the attenuated metric may be singular.

I still think it should be possible to use the attenuated metric to accomplish fitting, if one handles the inverse appropriately. The pivoted Cholesky scheme implemented in `orthog.cc` has no problems with singular basis sets.

Reply:
This brings us back to the original problem that using the inverted coulomb metric would require completely reformulating the blocking scheme for the case in which the doubly contracted eri tensors don't fit in memory. If we had known the efficacy of the code earlier, then I would have been happy to do it, but as of now, it's not a project I can undertake.

Reply:
Yeah. Just waiting on reviews from core devs